File: Firmware/RP2040/src/UserSettings/TriggerSettings.cpp
#include "UserSettings/TriggerSettings.h"

bool TriggerSettings::is_same(const TriggerSettingsRaw& raw) const
{
    return  dz_inner == Fix16(raw.dz_inner) &&
            dz_outer == Fix16(raw.dz_outer) &&
            anti_dz_inner == Fix16(raw.anti_dz_inner) &&
            anti_dz_outer == Fix16(raw.anti_dz_outer) &&
            curve == Fix16(raw.curve);
}

void TriggerSettings::set_from_raw(const TriggerSettingsRaw& raw)
{
    dz_inner = Fix16(raw.dz_inner);
    dz_outer = Fix16(raw.dz_outer);
    anti_dz_inner = Fix16(raw.anti_dz_inner);
    anti_dz_outer = Fix16(raw.anti_dz_outer);
    curve = Fix16(raw.curve);
}

File: Firmware/pico/web/modules/trigger/triggerSettings.js
export class TriggerSettings {
    static SETTINGS = Object.freeze([
        { type: "float", key: "dzInner",     min: 0.0, max: 0.5, def: 0.0, label: "Deadzone Inner", tip: "" },
        { type: "float", key: "dzOuter",     min: 0.5, max: 1.0, def: 1.0, label: "Deadzone Outer", tip: "" },
        { type: "float", key: "antiDzInner", min: 0.0, max: 0.5, def: 0.0, label: "Anti-Deadzone Inner", tip: "" },
        { type: "float", key: "antiDzOuter", min: 0.5, max: 1.0, def: 1.0, label: "Anti-Deadzone Outer", tip: "" },
        { type: "float", key: 'curve',       min: 0.3, max: 3.0, def: 1.0, label: "Curve", tip: "" },
    ]);

    constructor() {
        this.resetAll();
    }

    resetAll() {
        TriggerSettings.SETTINGS.forEach(setting => {
            this[setting.key] = setting.def;
        });
    }

    getDefaultValue(key) {
        const setting = TriggerSettings.SETTINGS.find(setting => setting.key === key);
        if (setting !== undefined) {
            return setting.def;
        }
        console.warn(`Setting not found: ${key}`);
        return null;
    }
}